### PR TITLE
Completed homework for lesson 2

### PR DESCRIPTION
**Pull request by mossbanay**
I've had a go at the data migrations and I seem to be having a problem where I still have the names of clients in what should be the client Foreign Key field of the project model.

Here's what the database looks like to help explain this.

```
sqlite> select * from entries_project;
id          name                 client_id
----------  -------------------  ---------------
1           Fascinating project  Creative People
2           Something boring     Lazy people
3           REALLY boring proje  Lazy people
```

### CODE REVIEW

File: timetracker/entries/models.py [hunk 1]
```diff
@@ -1,25 +1,33 @@
...
+class WorkPeriod(models.Model):
     start = models.DateTimeField(default=timezone.now)
     stop = models.DateTimeField(blank=True, null=True)
+
+    def __str__(self):
+        return '[{} - {}]'.format(self.start, self.stop)
+
+class Entry(models.Model):
     project = models.ForeignKey('Project')
+    workPeriod = models.ForeignKey('WorkPeriod', null=True)
```
Comment (sesh) [inline timetracker/entries/models.py]: In Python we'd normally use `work_period` as the attribute name here. CamelCasing is normally reserved for class names.

File: timetracker/entries/models.py [hunk 2]
```diff
@@ -1,25 +1,33 @@
...
+    client = models.ForeignKey('Client')
     name = models.CharField(max_length=200)
 
     def __str__(self):
         return '<{}> {}'.format(self.client, self.name)
 
-
-# Create your models here.
-class Entry(models.Model):
+class WorkPeriod(models.Model):
```
Comment (funkybob) [inline timetracker/entries/models.py:17]: Whilst I can see where your reasoning has gone here, separating WorkPeriod here doesn't really fit.

The reason is that the start/end time are specific to the Entry, and if two Entries share a start or end time it's purely coincidental, not meaningful.

File: timetracker/entries/migrations/0003_auto_20150726_0653.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+
+def retain_clients(apps, schema_editor):
+    Project = apps.get_model("entries", "Project")
+    Client = apps.get_model("entries", "Client")
+    for project in Project.objects.all():
+        # Use get_or_create to ensure that clients are duplicated
+        client, created = Client.objects.get_or_create(name=project.client)
+        if created:
+            client.save()
```
Comment (sesh) [inline timetracker/entries/migrations/0003_auto_20150726_0653.py]: The `get_or_create` function will save the object by default, so saving isn't required here.

File: timetracker/entries/migrations/0003_auto_20150726_0653.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+def retain_work_periods(apps, schema_editor):
+    Entry = apps.get_model("entries", "Entry")
+    WorkPeriod = apps.get_model("entries", "WorkPeriod")
+    for entry in Entry.objects.all():
+        workPeriod = WorkPeriod(start=entry.start, stop=entry.stop)
```
Comment (sesh) [inline timetracker/entries/migrations/0003_auto_20150726_0653.py:19]: The fact that at this point you're not using `get_or_create` to check if there is an existing `WorkPeriod` with the same time highlights what @funkybob is saying below.

These times are specific to a single `Entry` and won't be shared.

**Comment by sesh** [comment]
Awesome work on the standard parts of the homework.

We'll go over the bonus point solution at the start of the class on Thursday.

**Comment by mossbanay** [comment]
Thanks for the feedback @sesh @funkybob !

**Pull request closed without merging**